exonum/src/runtime/mod.rs

```diff
@@ -198,37 +198,42 @@ pub trait Runtime: Send + Debug + 'static {
     /// * Ensure that the deployed artifact has the following information, even if it is empty.
     fn artifact_protobuf_spec(&self, id: &ArtifactId) -> Option<ArtifactProtobufSpec>;
 
-    /// Restart previously added service instance with the given specification.
+    /// Runs the constructor of a new service instance with the given specification
+    /// and initial configuration.
+    ///
+    /// The service is not guaranteed to be added to the runtime at this point, so the runtime
+    /// may freely discard the instantiated service instance after completing this method. To
+    /// add the service to the runtime permanently, `add_service()` is called.
     ///
     /// # Policy on Panics
     ///
     /// * Catch each kind of panics except for `FatalError` and convert
     /// them into `ExecutionError`.
-    /// * If panic occurs, the runtime must ensure that it is in a consistent state.
-    fn restart_service(&mut self, spec: &InstanceSpec) -> Result<(), ExecutionError>;
+    fn start_adding_service(
```

Comment:
`start_adding_service` doesn't seem like a really good name for the method. I'd suggest `try_adding_service`.

Reply:
IMO, `try_adding_service` implies that adding is complete if the "normal" result is returned. This is not the case here; a call to `start_adding_service()` is not necessarily followed by adding the service "permanently" to the runtime.